library/blocks/in-world-recipes/src/main/java/org/quiltmc/qsl/inworldrecipes/api/InWorldRecipe.java

```diff
@@ -0,0 +1,22 @@
+package org.quiltmc.qsl.inworldrecipes.api;
+
+import net.minecraft.block.Block;
+import net.minecraft.item.ItemUsageContext;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Represents an <em>in-world recipe</em> - a recipe that can be performed on a specific block type in the world.
+ */
+public interface InWorldRecipe {
+	/**
+	 * {@return the specific block type required for this recipe}
+	 */
+	@NotNull Block targetBlock();
```

Comment:
Could this use a `Predicate<BlockState>` rather than just a single Block filter? Some may want to filter by block type (as vanilla recipes do) or filter by a specific block state.

As for the data driven part, the predicate can be specified similar to the way loot tables work
(Example, Revision 2)
```json
{
    "predicates": [
        {
            "type": "quilt:block_state",
            "block": "minecraft:oak_slab",
            "properties": {
                "half": "lower"
            }
        },
        {
            "type": "quilt:block_tag",
            "tag": "minecraft:flowers"
        },
        {
            "type": "custom:custom",
            "custom_condition_data": [ "foo", "bar" ]
        }
    ],
    "result": {
        "block": "minecraft:birch_slab",
        "properties": {}
    }
}
```

Reply:
This can be done by checking the `BlockState` in the `tryPerform` method before actually doing anything.

Reply:
What if you want to use a block tag as an input?

Reply:
...good point. Maybe I should just remove `targetBlock()` entirely.

Reply:
> ...good point. Maybe I should just remove `targetBlock()` entirely.

Couldn't you just make it optional? So that if a recipe will only work for one specific block, that can be optimized, but all other recipes will be run every time?

Reply:
> This can be done by checking the `BlockState` in the `tryPerform` method before actually doing anything.
As Will said, now you lose the flexibility of allowing multiple block types via tags (though my json example wouldn't work then)

Reply:
> Couldn't you just make it optional? So that if a recipe will only work for one specific block, that can be optimized, but all other recipes will be run every time?

That could work, though I'd need registry events to actually implement that optimization.

Reply:
```suggestion
	boolean matches(BlockState state);
```

the most basic impl would just be:
```java
if (this.block != null) {
	return state.getBlock() == this.block;
} else if (this.tag != null) {
	return tag.contains(state.getBlock());
}
```
but it can be extended to the other blockstate predicate stuff mentioned earlier

Reply:
> the most basic impl would just be:
> 
> ```java
> if (this.block != null) {
> 	return state.getBlock() == this.block;
> } else if (this.tag != null) {
> 	return tag.contains(state.getBlock());
> } else return false;
> ```
> 
> but it can be extended to the other blockstate predicate stuff mentioned earlier

You can just use `tryPerform` for this, instead of adding a new `matches` method.

Reply:
Ended up changing `Block targetBlock()` to `Set<Block> targetBlocks()`, I think this resolves the issue nicely (as I said, block state can be checked in `tryPerform`).